Content.Server/Anomaly/Effects/TileAnomalySystem.cs

```diff
@@ -17,25 +19,92 @@ public sealed class TileAnomalySystem : EntitySystem
+
+        SubscribeLocalEvent<TileSpawnAnomalyComponent, AnomalyPulseEvent>(OnPulse);
+        SubscribeLocalEvent<TileSpawnAnomalyComponent, AnomalySupercriticalEvent>(OnSupercritical);
+        SubscribeLocalEvent<TileSpawnAnomalyComponent, AnomalyStabilityChangedEvent>(OnStabilityChanged);
+        SubscribeLocalEvent<TileSpawnAnomalyComponent, AnomalySeverityChangedEvent>(OnSeverityChanged);
     }
 
-    private void OnSeverityChanged(EntityUid uid, TileSpawnAnomalyComponent component, ref AnomalyStabilityChangedEvent args)
+    private void OnPulse(Entity<TileSpawnAnomalyComponent> component, ref AnomalyPulseEvent args)
+    {
+        foreach (var entry in component.Comp.Entries)
+        {
+            if (!entry.SpawnOnPulse)
+                continue;
+
+            SpawnTiles(component, entry);
+        }
+    }
+    private void OnSupercritical(Entity<TileSpawnAnomalyComponent> component, ref AnomalySupercriticalEvent args)
     {
-        var xform = Transform(uid);
-        if (!_map.TryGetGrid(xform.GridUid, out var grid))
+        foreach (var entry in component.Comp.Entries)
+        {
+            if (!entry.SpawnOnSuperCritical)
+                continue;
+
+            SpawnTiles(component, entry);
+        }
+    }
+
+    private void OnStabilityChanged(Entity<TileSpawnAnomalyComponent> component, ref AnomalyStabilityChangedEvent args)
+    {
+        foreach (var entry in component.Comp.Entries)
+        {
+            if (!entry.SpawnOnStabilityChanged)
+                continue;
+
+            SpawnTiles(component, entry);
+        }
+    }
+
+    private void OnSeverityChanged(Entity<TileSpawnAnomalyComponent> component, ref AnomalySeverityChangedEvent args)
+    {
+        foreach (var entry in component.Comp.Entries)
+        {
+            if (!entry.SpawnOnSeverityChanged)
+                continue;
+
+            SpawnTiles(component, entry);
+        }
+    }
+
+    //TheShuEd:
+    //I know it's a shitcode! I didn't write it! I just restructured the functions
+    private void SpawnTiles(Entity<TileSpawnAnomalyComponent> component, TileSpawnSettingsEntry entry)
+    {
+        var xform = Transform(component.Owner);
+        if (!TryComp<MapGridComponent>(xform.GridUid, out var grid))
             return;
 
-        var radius = component.SpawnRange * args.Stability;
-        var fleshTile = (ContentTileDefinition) _tiledef[component.FloorTileId];
+        var amount = (int) (MathHelper.Lerp(entry.MinAmount, entry.MaxAmount, 1) + 0.5f);
+
         var localpos = xform.Coordinates.Position;
         var tilerefs = grid.GetLocalTilesIntersecting(
-            new Box2(localpos + new Vector2(-radius, -radius), localpos + new Vector2(radius, radius)));
+            new Box2(localpos + new Vector2(-entry.MaxRange, -entry.MaxRange), localpos + new Vector2(entry.MaxRange, entry.MaxRange))).ToArray();
+
+        if (tilerefs.Length == 0)
+            return;
+
+        _random.Shuffle(tilerefs);
```

Comment:
Don't copypaste between systems make it reusable.

Reply:
I don't know the best way to do this. It was copying a lot of code before refactor, so maybe it should be like this for now.